### PR TITLE
[core] Separate index cache and data cache

### LINKED ISSUE #4437: [Feature] Separate index cache and data cache in look up cache
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Motivation

Use a separate cache to store the index, such as bloomfilter & block index.

#3827

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Aitozi**
### Purpose

Linked issue: close #4437

### CODE REVIEW

File: docs/layouts/shortcodes/generated/core_configuration.html [hunk 1]
```diff
@@ -423,12 +417,24 @@
...
+            <td>Max memory size for lookup data cache.</td>
+        </tr>
         <tr>
             <td><h5>lookup.hash-load-factor</h5></td>
             <td style="word-wrap: break-word;">0.75</td>
             <td>Float</td>
             <td>The index load factor for lookup.</td>
         </tr>
+        <tr>
+            <td><h5>lookup.index-cache-max-memory-size</h5></td>
```
Comment (JingsongLi) [inline docs/layouts/shortcodes/generated/core_configuration.html:433]: See https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/deployment/config/#rocksdb-state-backend

Can we just introduce an option like `state.backend.rocksdb.memory.high-prio-pool-ratio`?

Comment (Aitozi) [reply docs/layouts/shortcodes/generated/core_configuration.html:433]: Make sense.

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 2]
```diff
@@ -915,6 +915,13 @@ public class CoreOptions implements Serializable {
                     .defaultValue(MemorySize.parse("256 mb"))
                     .withDescription("Max memory size for lookup cache.");
 
+    public static final ConfigOption<Double> LOOKUP_CACHE_HIGH_PRIO_POOL_RATIO =
+            key("lookup.cache.high-prio-pool-ratio")
```
Comment (wwj6591812) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:919]: I think lookup.cache.high-priority-pool-ratio may be better?

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java [hunk 3]
```diff
@@ -29,31 +33,63 @@
...
+    public CacheManager(MemorySize dataMaxMemorySize, double highPrioPoolRatio) {
+        this(Cache.CacheType.GUAVA, dataMaxMemorySize, highPrioPoolRatio);
     }
 
-    public CacheManager(Cache.CacheType cacheType, MemorySize maxMemorySize) {
-        this.cache = CacheBuilder.newBuilder(cacheType).maximumWeight(maxMemorySize).build();
+    public CacheManager(
+            Cache.CacheType cacheType, MemorySize maxMemorySize, double highPrioPoolRatio) {
+        Preconditions.checkArgument(
+                highPrioPoolRatio >= 0 && highPrioPoolRatio < 1,
```
Comment (wwj6591812) [inline paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java:61]: Do we need to pay attention to the accuracy of floating-point numbers here, which may lead to judgment errors?

File: paimon-common/src/main/java/org/apache/paimon/lookup/sort/BlockCache.java [hunk 4]
```diff
@@ -61,9 +60,9 @@ private byte[] readFrom(long offset, int length) throws IOException {
     }
 
     public MemorySegment getBlock(
-            long position, int length, Function<byte[], byte[]> decompressFunc) {
+            long position, int length, Function<byte[], byte[]> decompressFunc, boolean index) {
```
Comment (wwj6591812) [inline paimon-common/src/main/java/org/apache/paimon/lookup/sort/BlockCache.java:63]: May be isIndex?

Comment (Aitozi) [reply paimon-common/src/main/java/org/apache/paimon/lookup/sort/BlockCache.java]: fixed

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java (hunk 2 above)
Comment (Aitozi) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java]: fixed

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java (hunk 3 above)
Comment (Aitozi) [reply paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java]: e, I think it's ok here

File: docs/layouts/shortcodes/generated/core_configuration.html [hunk 5]
```diff
@@ -423,6 +423,12 @@
             <td>Double</td>
             <td>Define the default false positive probability for lookup cache bloom filters.</td>
         </tr>
+        <tr>
+            <td><h5>lookup.cache.high-priority-pool-ratio</h5></td>
+            <td style="word-wrap: break-word;">0.25</td>
```
Comment (JingsongLi) [inline docs/layouts/shortcodes/generated/core_configuration.html:428]: What is the source of this default value? 0.1 or 0.25?

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheKey.java [hunk 6]
```diff
@@ -24,25 +24,31 @@
 /** Key for cache manager. */
 public interface CacheKey {
 
-    static CacheKey forPosition(RandomAccessFile file, long position, int length) {
-        return new PositionCacheKey(file, position, length);
+    static CacheKey forPosition(RandomAccessFile file, long position, int length, boolean isIndex) {
+        return new PositionCacheKey(file, position, length, isIndex);
     }
 
     static CacheKey forPageIndex(RandomAccessFile file, int pageSize, int pageIndex) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/io/cache/CacheKey.java:31]: Can we remove this method? The invoker should always set the `isIndex`.

File: docs/layouts/shortcodes/generated/core_configuration.html (hunk 5 above)
Comment (Aitozi) [reply docs/layouts/shortcodes/generated/core_configuration.html:428]: if 0.1, the size is 25MB, 0.25 is 64MB. When I test, the bloom filter may occupy 5-10MB. So, I give a bigger (64MB) default index cache here.

**Comment by JingsongLi** [comment]
Hi @Aitozi , consider modifying old hash store, can you test it in your production env? I am not sure if this modification is good for old hash store.

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java (hunk 3 above)
Comment (wwj6591812) [reply paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java]: ok

**Comment by Aitozi** [comment]
> Hi @Aitozi , consider modifying old hash store, can you test it in your production env? I am not sure if this modification is good for old hash store.

Sure, will test it first

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**